distribution/pull_v2.go

```diff
@@ -175,96 +180,106 @@ func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progre
-		}
-		return nil, 0, retryOnError(err)
-	}
-
-	if offset != 0 {
-		_, err := layerDownload.Seek(offset, os.SEEK_SET)
+	if ld.verifier == nil || !ld.verifier.Verified() {
+		progress.Update(progressOutput, ld.ID(), "Connecting")
+		layerDownload, err := ld.open(ctx)
 		if err != nil {
-			if err := ld.truncateDownloadFile(); err != nil {
+			logrus.Errorf("Error initiating layer download: %v", err)
+			if err == distribution.ErrBlobUnknown {
 				return nil, 0, xfer.DoNotRetry{Err: err}
 			}
-			return nil, 0, err
+			return nil, 0, retryOnError(err)
 		}
-	}
-	size, err := layerDownload.Seek(0, os.SEEK_END)
-	if err != nil {
-		// Seek failed, perhaps because there was no Content-Length
-		// header. This shouldn't fail the download, because we can
-		// still continue without a progress bar.
-		size = 0
-	} else {
-		if size != 0 && offset > size {
-			logrus.Debug("Partial download is larger than full blob. Starting over")
-			offset = 0
-			if err := ld.truncateDownloadFile(); err != nil {
-				return nil, 0, xfer.DoNotRetry{Err: err}
+
+		if offset != 0 {
+			_, err := layerDownload.Seek(offset, os.SEEK_SET)
+			if err != nil {
+				if err := ld.truncateDownloadFile(); err != nil {
+					return nil, 0, xfer.DoNotRetry{Err: err}
+				}
+				return nil, 0, err
 			}
 		}
-
-		// Restore the seek offset either at the beginning of the
-		// stream, or just after the last byte we have from previous
-		// attempts.
-		_, err = layerDownload.Seek(offset, os.SEEK_SET)
+		size, err := layerDownload.Seek(0, os.SEEK_END)
 		if err != nil {
-			return nil, 0, err
+			logrus.Debugf("failed to determine download size for %s: %s", ld.ID(), err)
+			// Seek failed, perhaps because there was no Content-Length
+			// header. This shouldn't fail the download, because we can
+			// still continue without a progress bar.
+			size = 0
+		} else {
+			if size != 0 && offset > size {
+				logrus.Debug("Partial download is larger than full blob. Starting over")
+				offset = 0
+				if err := ld.truncateDownloadFile(); err != nil {
+					return nil, 0, xfer.DoNotRetry{Err: err}
+				}
+			}
+
+			// Restore the seek offset either at the beginning of the
+			// stream, or just after the last byte we have from previous
+			// attempts.
+			_, err = layerDownload.Seek(offset, os.SEEK_SET)
+			if err != nil {
+				return nil, 0, err
+			}
 		}
-	}
 
-	reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, layerDownload), progressOutput, size-offset, ld.ID(), "Downloading")
-	defer reader.Close()
+		reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, layerDownload), progressOutput, offset, size, ld.ID(), "Downloading")
+		defer reader.Close()
 
-	if ld.verifier == nil {
-		ld.verifier, err = digest.NewDigestVerifier(ld.digest)
-		if err != nil {
-			return nil, 0, xfer.DoNotRetry{Err: err}
+		if ld.verifier == nil {
+			logrus.Debugf("new verifier for %s", ld.ID())
+			ld.verifier, err = digest.NewDigestVerifier(ld.digest)
```

Comment:
How is this verifier populated with the partial digest state?

We either have to read data from disk or use resumable hashes. I am having trouble understanding if we consume this data in some way.